File: Packages/Flip54Storage/Sources/Flip54Storage/ActiveSession.swift
import Foundation
import Flip54Core

public struct ActiveSession: Codable, Hashable, Sendable {
    public let id: UUID
    public let deckId: String
    public let startedAt: Date
    public let equipment: Equipment
    public let difficulty: Difficulty

    public var drawPile: [Card]
    public var discardPile: [Card]
    public var currentCard: Card?

    public var totalRepsCompleted: Int
    public var holdSecondsCompleted: Int
    public var skipCount: Int
    public var cardsCompleted: Int
    public var repsBySuit: [Suit: Int]

    public var holdStartedAt: Date?
    public var holdElapsed: TimeInterval

    public var pauseStartedAt: Date?
    public var totalPausedDuration: TimeInterval

    public var cardsRemaining: Int {
        drawPile.count + (currentCard != nil ? 1 : 0)
    }

    public var isComplete: Bool {
        drawPile.isEmpty && currentCard == nil
    }

    // MARK: - Factory

    public static func start(
        deck: [Card],
        deckId: String,
        equipment: Equipment,
        difficulty: Difficulty,
        rng: inout some RandomNumberGenerator
    ) -> ActiveSession {
        var shuffled = deck
        shuffled.shuffle(using: &rng)
        return ActiveSession(
            id: UUID(),
            deckId: deckId,
            startedAt: Date(),
            equipment: equipment,
            difficulty: difficulty,
            drawPile: shuffled,
            discardPile: [],
            currentCard: nil,
            totalRepsCompleted: 0,
            holdSecondsCompleted: 0,
            skipCount: 0,
            cardsCompleted: 0,
            repsBySuit: [:],
            holdStartedAt: nil,
            holdElapsed: 0,
            pauseStartedAt: nil,
            totalPausedDuration: 0
        )
    }

    // MARK: - Mutations

    public mutating func flipNextCard() {
        guard currentCard == nil, !drawPile.isEmpty else { return }
        currentCard = drawPile.removeLast()
        holdElapsed = 0
        holdStartedAt = nil
    }

    public mutating func startHold(at date: Date) {
        guard let card = currentCard, card.isAce else { return }
        holdStartedAt = date
    }

    public mutating func completeCurrentCard(reps: Int?, holdSeconds: Int?) {
        guard let card = currentCard else { return }
        if let reps {
            totalRepsCompleted += reps
            if case .standard(let suit, _) = card {
                repsBySuit[suit, default: 0] += reps
            }
        }
        if let secs = holdSeconds {
            holdSecondsCompleted += secs
        }
        discardPile.append(card)
        cardsCompleted += 1
        currentCard = nil
        holdStartedAt = nil
        holdElapsed = 0
    }

    public mutating func skipCurrentCard() {
        guard let card = currentCard else { return }
        drawPile.insert(card, at: 0)  // bottom of draw pile (index 0 = bottom)
        skipCount += 1
        currentCard = nil
        holdStartedAt = nil
        holdElapsed = 0
    }

    public mutating func pause(at date: Date) {
        guard pauseStartedAt == nil else { return }
        pauseStartedAt = date
        if holdStartedAt != nil {
            holdElapsed += date.timeIntervalSince(holdStartedAt!)
            holdStartedAt = nil
        }
    }

    public mutating func resume(at date: Date) {
        guard let pauseStart = pauseStartedAt else { return }
        totalPausedDuration += date.timeIntervalSince(pauseStart)
        pauseStartedAt = nil
    }
}

File: android/storage/src/main/kotlin/com/flip54/storage/ActiveSession.kt
package com.flip54.storage

import com.flip54.core.*
import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class ActiveSession(
    val id: String = UUID.randomUUID().toString(),
    val deckId: String,
    val startedAtMs: Long,
    val equipment: Equipment,
    val difficulty: Difficulty,
    var drawPile: List<Card>,
    var discardPile: List<Card> = emptyList(),
    var currentCard: Card? = null,
    var totalRepsCompleted: Int = 0,
    var holdSecondsCompleted: Int = 0,
    var skipCount: Int = 0,
    var cardsCompleted: Int = 0,
    var heartsReps: Int = 0,
    var spadesReps: Int = 0,
    var clubsReps: Int = 0,
    var diamondsReps: Int = 0,
    var holdStartedAtMs: Long? = null,
    var holdElapsedMs: Long = 0,
    var pauseStartedAtMs: Long? = null,
    var totalPausedDurationMs: Long = 0
) {
    val cardsRemaining: Int get() = drawPile.size + (if (currentCard != null) 1 else 0)

    val deckSize: Int get() = drawPile.size + (if (currentCard != null) 1 else 0) + cardsCompleted

    val isComplete: Boolean get() = drawPile.isEmpty() && currentCard == null

    val repsBySuit: Map<Suit, Int> get() = mapOf(
        Suit.HEARTS   to heartsReps,
        Suit.SPADES   to spadesReps,
        Suit.CLUBS    to clubsReps,
        Suit.DIAMONDS to diamondsReps
    )

    fun flipNextCard(): ActiveSession {
        if (currentCard != null || drawPile.isEmpty()) return this
        val next = drawPile.last()
        return copy(
            drawPile = drawPile.dropLast(1),
            currentCard = next,
            holdElapsedMs = 0,
            holdStartedAtMs = null
        )
    }

    fun startHold(nowMs: Long): ActiveSession {
        val card = currentCard ?: return this
        if (!card.isAce) return this
        return copy(holdStartedAtMs = nowMs)
    }

    fun completeCurrentCard(reps: Int?, holdSeconds: Int?, nowMs: Long): ActiveSession {
        val card = currentCard ?: return this
        var s = this
        if (reps != null) {
            s = s.copy(totalRepsCompleted = s.totalRepsCompleted + reps)
            if (card is Card.Standard) {
                s = when (card.suit) {
                    Suit.HEARTS   -> s.copy(heartsReps = s.heartsReps + reps)
                    Suit.SPADES   -> s.copy(spadesReps = s.spadesReps + reps)
                    Suit.CLUBS    -> s.copy(clubsReps = s.clubsReps + reps)
                    Suit.DIAMONDS -> s.copy(diamondsReps = s.diamondsReps + reps)
                }
            }
        }
        if (holdSeconds != null) {
            s = s.copy(holdSecondsCompleted = s.holdSecondsCompleted + holdSeconds)
        }
        return s.copy(
            discardPile = s.discardPile + card,
            cardsCompleted = s.cardsCompleted + 1,
            currentCard = null,
            holdStartedAtMs = null,
            holdElapsedMs = 0
        )
    }

    fun skipCurrentCard(): ActiveSession {
        val card = currentCard ?: return this
        return copy(
            drawPile = listOf(card) + drawPile,
            skipCount = skipCount + 1,
            currentCard = null,
            holdStartedAtMs = null,
            holdElapsedMs = 0
        )
    }

    fun pause(nowMs: Long): ActiveSession {
        if (pauseStartedAtMs != null) return this
        val holdStart = holdStartedAtMs
        val extraElapsed = if (holdStart != null) nowMs - holdStart else 0L
        return copy(
            pauseStartedAtMs = nowMs,
            holdElapsedMs = holdElapsedMs + extraElapsed,
            holdStartedAtMs = null
        )
    }

    fun resume(nowMs: Long): ActiveSession {
        val pauseStart = pauseStartedAtMs ?: return this
        return copy(
            totalPausedDurationMs = totalPausedDurationMs + (nowMs - pauseStart),
            pauseStartedAtMs = null
        )
    }

    companion object {
        fun start(
            deck: List<Card>,
            deckId: String,
            equipment: Equipment,
            difficulty: Difficulty,
            nowMs: Long = System.currentTimeMillis()
        ): ActiveSession = ActiveSession(
            deckId = deckId,
            startedAtMs = nowMs,
            equipment = equipment,
            difficulty = difficulty,
            drawPile = deck.shuffled()
        )
    }
}
